File: programmers/level_3/가장먼노드.js
function solution(n, edge) {
  var matrix = Array.from(Array(n + 1), () => Array());
  var visit = new Array(n + 1).fill(false);
  var minDis = new Array(n + 1).fill(Number.MAX_SAFE_INTEGER);

  // 인접 리스트
  for (var e of edge) {
    matrix[e[0]].push(e[1]);
    matrix[e[1]].push(e[0]);
  }

  var q = [];
  q.push(1);

  minDis[0] = 0;
  minDis[1] = 0;
  visit[1] = true;
  while (q.length) {
    var from = q.shift();
    matrix[from].forEach((to) => {
      // 방문 체크
      if (!visit[to]) {
        visit[to] = true;
        q.push(to);
      }
      // 거리 저장
      minDis[to] =
        minDis[to] > minDis[from] + 1 ? minDis[from] + 1 : minDis[to];
    });
  }
  var max = Math.max(...minDis);
  var answer = minDis.filter((d) => d === max);
  return answer.length;
}

File: programmers/level_3/가장먼노드.py
# 그래프의 최단거리를 찾기 위해서 bfs를 활용한다.
# 방문체크배열 visit과 거리를 담아둘 distance를 초기화한다.
# 가장 멀리 떨어진 노드는 bfs가 제일 처음 방문한 곳에 [이전노드의 거리 + 1]을 더해주고 다시방문하지 않으면된다.
from collections import deque

def solution(n, edge):
    visit = [False] * (n + 1)
    distance = [0] * (n + 1)
    G = [[] for _ in range(n + 1)]
    for node in edge:
        G[node[0]].append(node[1])
        G[node[1]].append(node[0])
        
    
    visit[1] = True
    q = deque()
    q.append(1)
    while q:
        v = q.popleft()
        for w in G[v]:
            if not visit[w]:
                visit[w] = True
                distance[w] = distance[v] + 1
                q.append(w)
    
    return distance.count(max(distance))